src/dbnode/encoding/m3tsz/iterator.go

```diff
@@ -83,6 +83,49 @@ func (it *readerIterator) Next() bool {
 		return false
 	}
 
+	if it.tsIterator.PrevTime != 0 && !it.tsIterator.TimeUnitChanged {
+		if it.intOptimized {
+			// This is the fast path to read the shortest encoding possible (001 binary) which
+			// represents an unchanged datapoint value and unchanged delta (deltaOfDelta = 0).
+			//  Breakdown of 001 opcode sequence:
+			// 0: TimeEncodingScheme.ZeroBucket().Opcode()
+			// 0: opcodeUpdate
+			// 1: opcodeRepeat
+			if it.peekBits(3) == 0b001 {
+				if it.err != nil {
+					return false
+				}
+
+				it.tsIterator.PrevTime += xtime.UnixNano(it.tsIterator.PrevTimeDelta)
+				it.curr.TimestampNanos = it.tsIterator.PrevTime
+				it.tsIterator.PrevAnt = nil
+
+				it.readBits(3) // consume the bits that were checked with peekBits
+
+				return it.hasNext()
+			}
+		} else { // non-int optimized
+			// This is the fast path to read the shortest encoding possible (00 binary) which
+			// represents an unchanged datapoint value and unchanged delta (deltaOfDelta = 0).
+			//  Breakdown of 00 opcode sequence:
+			// 0: TimeEncodingScheme.ZeroBucket().Opcode()
+			// 0: opcodeZeroValueXOR
+			if it.peekBits(2) == 0b00 {
+				if it.err != nil {
+					return false
+				}
+
+				it.tsIterator.PrevTime += xtime.UnixNano(it.tsIterator.PrevTimeDelta)
+				it.curr.TimestampNanos = it.tsIterator.PrevTime
+				it.tsIterator.PrevAnt = nil
```

Comment:
`FloatEncoderAndIterator` seems to also be setting `PrevXor = 0` on `opcodeZeroValueXOR`

https://github.com/m3db/m3/blob/87f37951bf2274b6d5560ec595ce508ac4944fc1/src/dbnode/encoding/m3tsz/float_encoder_iterator.go#L123-L126

Reply:
Good catch. Funny thing though, I believe, this would be impossible to reproduce even in theory :)

Reply:
I mean, with using the encoder. Probably possible to craft such input manually.